### PR TITLE
add by_authorized withdraw objects to get_full_accounts

### PR DESCRIPTION
**Pull request by oxarbitrage**
For https://github.com/bitshares/bitshares-core/issues/1654

As `from` and `to` were added in `get_full_accounts` for htlc objects i think the best is to resolve the ticket in the same way for withdraws.

Should not affect current client implementations.

### CODE REVIEW

**Comment by abitmore** [comment]
See comment in the issue. I don't think we should add it without checking result size.

**Comment by oxarbitrage** [comment]
Added 2 commits:

- In https://github.com/bitshares/bitshares-core/pull/1749/commits/cab48aef565104688d04a0810a4907b38473b3b3 `api-limit-get-full-accounts` is added and also a bug in `api-limit-get-htlc-by` is fixed.
- In https://github.com/bitshares/bitshares-core/pull/1749/commits/ee0df87ec8bca801a1361990e3e3f1e456e42349 limits are applied to `get_full_accounts` list results.

One problem i see is that the in the `for_each` lists, i cant exit the loop early. After limit, no more inserts will be done to the results but loop is still iterating. I cant find a clean way to resolve this problem, not sure either if it is acceptable or "good enough" as it is either.

**Comment by pmconrad** [comment]
How about using a normal for loop and breaking out if the limit is reached? :-)

File: libraries/app/database_api.cpp [hunk 1]
```diff
@@ -921,71 +921,98 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
+            [&acnt, api_limit_get_full_accounts] (const htlc_object& htlc) {
+               if(acnt.htlcs.size() < api_limit_get_full_accounts)
+                  acnt.htlcs.emplace_back(htlc);
             });
       auto htlc_to_range = _db.get_index_type<htlc_index>().indices().get<by_to_id>().equal_range(account->id);
       std::for_each(htlc_to_range.first, htlc_to_range.second,
-            [&acnt] (const htlc_object& htlc) {
-               if ( std::find(acnt.htlcs.begin(), acnt.htlcs.end(), htlc) == acnt.htlcs.end() )
+            [&acnt, api_limit_get_full_accounts] (const htlc_object& htlc) {
+               if ((std::find(acnt.htlcs.begin(), acnt.htlcs.end(), htlc) == acnt.htlcs.end()) or
```
Comment (pmconrad) [inline libraries/app/database_api.cpp]: Using `find` inside the `for_each` loop leads to quadratic complexity. Since this is only useful for the abnormal case of an htlc with `from == to` I suggest to drop the `find` entirely.

File: libraries/app/application.cpp [hunk 2]
```diff
@@ -1020,6 +1023,10 @@ void application::set_program_options(boost::program_options::options_descriptio
           "For asset_api::get_asset_holders to set its default limit value as 100")
 		   ("api-limit-get-key-references",boost::program_options::value<uint64_t>()->default_value(100),
 		    "For database_api_impl::get_key_references to set its default limit value as 100")
+         ("api-limit-get-htlc-by",boost::program_options::value<uint64_t>()->default_value(100),
+          "For database_api_impl::get_htlc_by_from and get_htlc_by_to to set its default limit value as 100")
+         ("api-limit-get-full-accounts",boost::program_options::value<uint64_t>()->default_value(100),
```
Comment (pmconrad) [inline libraries/app/application.cpp]: IMO the option name is confusing. The other limits apply to the number of list entries returned by the call. Here, the number of accounts is unlimited, but the setting limits the number of entries inside each account's sub-lists.

IMO we should limit the number of accounts that can be queried in one go as well. This is an easy DOS vector.

Comment (oxarbitrage) [reply libraries/app/application.cpp]: fixed this by having 2 limit optiosn for get_full_accounts: `api-limit-get-full-accounts` for the account vector and `api-limit-get-full-accounts-lists` for the list results.

Comment (pmconrad) [reply libraries/app/application.cpp]: Thanks. @sschiessl-bcp does the UI call `get_full_accounts` with many accounts at once? How many typically?

Comment (oxarbitrage) [reply libraries/app/application.cpp]: Actually the default is currently 10 accounts but i forgot to change the 100 here, https://github.com/bitshares/bitshares-core/blob/3aba3f76d47e6e4b20707c69a51de43d5dc3f84b/libraries/app/include/graphene/app/application.hpp#L51
Still, will wait for @sschiessl-bcp response before fixing this and changing in the other places if needed.

File: libraries/app/database_api.cpp [hunk 3]
```diff
@@ -921,71 +921,127 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
+         }
       }
 
-
       // Add the account's balances
       const auto& balances = _db.get_index_type< primary_index< account_balance_index > >().get_secondary_index< balances_by_account_index >().get_account_balances( account->id );
       for( const auto balance : balances )
+      {
          acnt.balances.emplace_back( *balance.second );
+         if(acnt.balances.size() >= api_limit_get_full_accounts) {
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: I think the check is ok there to guarantee the correct number of output results. 

What can be done and i did here https://github.com/bitshares/bitshares-core/pull/1749/commits/445c6d4f7da16883a7bab304e05cb116d580de6e is to calculate of there is additional data before entering each loop.

File: libraries/app/database_api.cpp [hunk 4]
```diff
@@ -921,71 +921,127 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
       auto  required_approvals_itr = proposals_by_account._account_to_proposals.find( account->id );
       if( required_approvals_itr != proposals_by_account._account_to_proposals.end() )
       {
-         acnt.proposals.reserve( required_approvals_itr->second.size() );
+         acnt.proposals.reserve( std::min(required_approvals_itr->second.size(), api_limit_get_full_accounts) );
          for( auto proposal_id : required_approvals_itr->second )
+         {
             acnt.proposals.push_back( proposal_id(_db) );
+            if(acnt.proposals.size() >= api_limit_get_full_accounts) {
+               acnt.more_data_available.proposals = true;
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: done at https://github.com/bitshares/bitshares-core/pull/1749/commits/2fd3db2c9d660c6ca51685f279b91989c107afa9

File: libraries/app/database_api.cpp [hunk 5]
```diff
@@ -921,71 +921,127 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
       auto htlc_from_range = _db.get_index_type<htlc_index>().indices().get<by_from_id>().equal_range(account->id);
-      std::for_each(htlc_from_range.first, htlc_from_range.second,
-            [&acnt] (const htlc_object& htlc) {
-               acnt.htlcs.emplace_back(htlc);
-            });
+      for(auto itr = htlc_from_range.first; itr != htlc_from_range.second; ++itr)
+      {
+         acnt.htlcs.emplace_back(*itr);
+         if(acnt.htlcs.size() >= api_limit_get_full_accounts) {
+            acnt.more_data_available.htlcs = true;
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: separated the lists and also in withdraws as it will suffer the same issue. https://github.com/bitshares/bitshares-core/pull/1749/commits/c20c3254dcc243ee15f6913446bf772dd8b71f81

File: libraries/app/database_api.cpp [hunk 6]
```diff
@@ -922,124 +925,124 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
-            if(acnt.proposals.size() >= api_limit_get_full_accounts) {
-               acnt.more_data_available.proposals = true;
+            if(acnt.proposals.size() >= api_limit_get_full_accounts_lists)
                break;
-            }
          }
       }
 
       // Add the account's balances
       const auto& balances = _db.get_index_type< primary_index< account_balance_index > >().get_secondary_index< balances_by_account_index >().get_account_balances( account->id );
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: While you're on it, can you please wrap this long line? Thanks.

File: libraries/app/database_api.cpp [hunk 7]
```diff
@@ -922,124 +925,124 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
-         if(acnt.balances.size() >= api_limit_get_full_accounts) {
-            acnt.more_data_available.balances = true;
+         if(acnt.balances.size() >= api_limit_get_full_accounts_lists)
             break;
-         }
       }
 
       // Add the account's vesting balances
       auto vesting_range = _db.get_index_type<vesting_balance_index>().indices().get<by_account>().equal_range(account->id);
+      if((unsigned)abs(distance(vesting_range.first, vesting_range.second)) > api_limit_get_full_accounts_lists)
```
Comment (abitmore) [inline libraries/app/database_api.cpp]: According to https://en.cppreference.com/w/cpp/iterator/distance, complexity of `distance()` should be linear for boost multi-index container which is implemented with red-black tree. IMHO we should avoid `instance()` call when possible. Checking size before pushing a new item to the list is cheaper when the distance is large, also not more expensive when distance is small.
```
for( itr = start; itr != end; ++itr )
{
  if( list_size >= max )
   {
      has_more = true;
      break;
   }
   insert_to_list;
}
```

File: libraries/app/include/graphene/app/database_api.hpp [hunk 8]
```diff
@@ -444,6 +444,14 @@ class database_api
        */
       vector<call_order_object> get_call_orders(const std::string& a, uint32_t limit)const;
 
+      /**
+       * @brief Get call orders from a given account
+       * @param account_name_or_id Account name or ID to get objects from
+       * @param limit Maximum number of objects to retrieve
+       * @return The call orders of the account
+       */
+      vector<call_order_object> get_call_orders_by_account(const std::string& account_name_or_id, uint32_t limit)const;
```
Comment (abitmore) [inline libraries/app/include/graphene/app/database_api.hpp]: Need a `start` or `lower_bound` parameter to be able to fetch next pages.

File: libraries/app/database_api.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (pmconrad) [reply libraries/app/database_api.cpp]: The number is correct, but the flag may be set incorrectly (i. e. flag will be set to true if total number == limit).

File: libraries/app/database_api.cpp [hunk 9]
```diff
@@ -2362,22 +2362,24 @@ vector<proposal_object> database_api::get_proposed_transactions( const std::stri
    return my->get_proposed_transactions( account_id_or_name );
 }
 
-/** TODO: add secondary index that will accelerate this process */
 vector<proposal_object> database_api_impl::get_proposed_transactions( const std::string account_id_or_name )const
 {
-   const auto& idx = _db.get_index_type<proposal_index>();
+   const auto& proposal_idx = _db.get_index_type<proposal_index>();
+   const auto& pidx = dynamic_cast<const base_primary_index&>(proposal_idx);
```
Comment (pmconrad) [inline libraries/app/database_api.cpp]: I think you can combine the two lines above into `... idx = _db.get_index_type< primary_index< proposal_index > >();`

File: libraries/app/application.cpp (hunk 2 above)
Comment (sschiessl-bcp) [reply libraries/app/application.cpp]: `bitsharesjs` only uses one account at all times, so does UI (at least up what I could find)

File: libraries/app/include/graphene/app/full_account.hpp [hunk 10]
```diff
@@ -40,8 +40,10 @@ namespace graphene { namespace app {
       bool settle_orders = false;
       bool proposals = false;
       bool assets = false;
-      bool withdraws = false;
-      bool htlcs = false;
+      bool withdraws_from = false;
+      bool withdraws_authorized = false;
```
Comment (pmconrad) [inline libraries/app/include/graphene/app/full_account.hpp]: Perhaps `withdraws_to` for consistency with htlc?

File: libraries/app/application.cpp (hunk 2 above)
Comment (sschiessl-bcp) [reply libraries/app/application.cpp]: Speaking of: How hard is it to add pagination, or limit to `get_key_references` and `get_account_references`?

There have been attacks on several accounts by creating dummy accounts and adding keys of the account to attack.

File: libraries/app/database_api.cpp (hunk 7 above)
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: ok, thank you for the detailed explication. i had used your code in all lists here https://github.com/bitshares/bitshares-core/pull/1749/commits/bda09a0989d8b5c030792b1dc6ecc12217b43868

File: libraries/app/database_api.cpp (hunk 9 above)
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: Thank you, also used the code to do the same in `get_full_accounts` at https://github.com/bitshares/bitshares-core/pull/1749/commits/66c0f40ae89130ce018ffb4c92a5d588d698c3d9

File: libraries/app/database_api.cpp (hunk 6 above)
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: done at https://github.com/bitshares/bitshares-core/pull/1749/commits/65d518021bd98250dcb928e29bae3b412072ed57

File: libraries/app/include/graphene/app/full_account.hpp (hunk 10 above)
Comment (oxarbitrage) [reply libraries/app/include/graphene/app/full_account.hpp]: changed at https://github.com/bitshares/bitshares-core/pull/1749/commits/65d518021bd98250dcb928e29bae3b412072ed57

File: libraries/app/include/graphene/app/database_api.hpp (hunk 8 above)
Comment (oxarbitrage) [reply libraries/app/include/graphene/app/database_api.hpp]: done at https://github.com/bitshares/bitshares-core/pull/1749/commits/38c3cef72d6379b50b2268fe908a84d467ca6b93

File: libraries/app/database_api.cpp [hunk 11]
```diff
@@ -921,71 +921,127 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
       auto asset_range = _db.get_index_type<asset_index>().indices().get<by_issuer>().equal_range(account->id);
-      std::for_each(asset_range.first, asset_range.second,
-                    [&acnt] (const asset_object& asset) {
-                       acnt.assets.emplace_back(asset.id);
-                    });
+      for(auto itr = asset_range.first; itr != asset_range.second; ++itr)
+      {
+         acnt.assets.emplace_back(itr->id);
+         if(acnt.assets.size() >= api_limit_get_full_accounts) {
+            acnt.more_data_available.assets = true;
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: `added at https://github.com/bitshares/bitshares-core/pull/1749/commits/ce709f9abc1ac31e920b6bbf85107dff9a0da3f6`

File: libraries/app/database_api.cpp [hunk 12]
```diff
@@ -921,71 +921,127 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
+      }
       auto settle_range = _db.get_index_type<force_settlement_index>().indices().get<by_account>().equal_range(account->id);
-      std::for_each(settle_range.first, settle_range.second,
-                    [&acnt] (const force_settlement_object& settle) {
-                       acnt.settle_orders.emplace_back(settle);
-                    });
+      for(auto itr = settle_range.first; itr != settle_range.second; ++itr)
+      {
+         acnt.settle_orders.emplace_back(*itr);
+         if(acnt.settle_orders.size() >= api_limit_get_full_accounts) {
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: added at https://github.com/bitshares/bitshares-core/pull/1749/commits/cd504f59c409bd3b9740435e9a94968791cf91a7

File: libraries/app/database_api.cpp [hunk 13]
```diff
@@ -921,71 +921,127 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
...
       auto call_range = _db.get_index_type<call_order_index>().indices().get<by_account>().equal_range(account->id);
-      std::for_each(call_range.first, call_range.second,
-                    [&acnt] (const call_order_object& call) {
-                       acnt.call_orders.emplace_back(call);
-                    });
+      for(auto itr = call_range.first; itr != call_range.second; ++itr)
+      {
+         acnt.call_orders.emplace_back(*itr);
+         if(acnt.call_orders.size() >= api_limit_get_full_accounts) {
+            acnt.more_data_available.call_orders = true;
```
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: added at https://github.com/bitshares/bitshares-core/pull/1749/commits/413b9cfe004bee5b57990bf7831f066f661f6101 and https://github.com/bitshares/bitshares-core/pull/1749/commits/38c3cef72d6379b50b2268fe908a84d467ca6b93

File: libraries/app/database_api.cpp [hunk 14]
```diff
@@ -928,6 +924,8 @@ std::map<std::string, full_account> database_api_impl::get_full_accounts( const
       uint64_t api_limit_get_full_accounts_lists = _app_options->api_limit_get_full_accounts_lists;
 
       // Add the account's proposals
+      const auto& proposal_idx = _db.get_index_type< primary_index< proposal_index > >();
```
Comment (pmconrad) [inline libraries/app/database_api.cpp]: Why did you move this inside the for loop?

File: libraries/chain/include/graphene/chain/market_object.hpp [hunk 15]
```diff
@@ -214,6 +215,12 @@ typedef multi_index_container<
             const_mem_fun< call_order_object, asset_id_type, &call_order_object::debt_type>
          >
       >,
+      ordered_unique< tag<by_account_and_id>,
```
Comment (pmconrad) [inline libraries/chain/include/graphene/chain/market_object.hpp]: Hm. We could avoid this additional index if we paginate on `asset_id`.
I think it's very unlikely that a new `call_order_object` appears while someone is cycling through the list, especially because only very few accounts will have more than one page of call orders.
@abitmore ?

File: libraries/app/include/graphene/app/database_api.hpp [hunk 16]
```diff
@@ -462,6 +462,16 @@ class database_api
        */
       vector<force_settlement_object> get_settle_orders(const std::string& a, uint32_t limit)const;
 
+      /**
+       * @brief Get forced settlement orders of a given account
+       * @param account_name_or_id Account name or ID to get objects from
+       * @param start Withdraw permission objects(1.4.X) before this ID will be skipped in results. Pagination purposes.
```
Comment (pmconrad) [inline libraries/app/include/graphene/app/database_api.hpp]: C&P

File: libraries/app/include/graphene/app/database_api.hpp [hunk 17]
```diff
@@ -423,6 +423,16 @@ class database_api
        */
       uint64_t get_asset_count()const;
 
+      /**
+       * @brief Get asset objects issued from a given account
+       * @param account_name_or_id Account name or ID to get objects from
+       * @param start Withdraw permission objects(1.3.X) before this ID will be skipped in results. Pagination purposes.
```
Comment (pmconrad) [inline libraries/app/include/graphene/app/database_api.hpp]: C&P

File: libraries/chain/include/graphene/chain/asset_object.hpp [hunk 18]
```diff
@@ -305,6 +306,12 @@ namespace graphene { namespace chain {
                 const_mem_fun<asset_object, bool, &asset_object::is_market_issued>,
                 member< object, object_id_type, &object::id >
             >
+         >,
+         ordered_unique< tag<by_issuer_and_id>,
```
Comment (pmconrad) [inline libraries/chain/include/graphene/chain/asset_object.hpp]: You can replace the `by_issuer` index with this one.

File: libraries/app/include/graphene/app/database_api.hpp [hunk 19]
```diff
@@ -444,6 +454,16 @@ class database_api
        */
       vector<call_order_object> get_call_orders(const std::string& a, uint32_t limit)const;
 
+      /**
+       * @brief Get call orders from a given account
+       * @param account_name_or_id Account name or ID to get objects from
+       * @param start Withdraw permission objects(1.8.X) before this ID will be skipped in results. Pagination purposes.
```
Comment (pmconrad) [inline libraries/app/include/graphene/app/database_api.hpp]: C&P

**Comment by pmconrad** [comment]
Please add test cases for the new API calls.

File: libraries/chain/include/graphene/chain/market_object.hpp (hunk 15 above)
Comment (abitmore) [reply libraries/chain/include/graphene/chain/market_object.hpp]: I agree it's good to reuse `by_account`.

File: libraries/app/database_api.cpp (hunk 14 above)
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: My mistake, i didn't saw the implication of the move correctly, my apologies. reverted at https://github.com/bitshares/bitshares-core/pull/1749/commits/ed53cb09fd677fa95d01cd3ad10b0d2fe0acd924

File: libraries/app/include/graphene/app/database_api.hpp (hunk 19 above)
Comment (oxarbitrage) [reply libraries/app/include/graphene/app/database_api.hpp]: fixed at https://github.com/bitshares/bitshares-core/pull/1749/commits/0968f95c2e8b761808bc3058ebfae44a21814dbc

File: libraries/app/include/graphene/app/database_api.hpp (hunk 16 above)
Comment (oxarbitrage) [reply libraries/app/include/graphene/app/database_api.hpp]: fixed at https://github.com/bitshares/bitshares-core/pull/1749/commits/0968f95c2e8b761808bc3058ebfae44a21814dbc

File: libraries/app/include/graphene/app/database_api.hpp (hunk 17 above)
Comment (oxarbitrage) [reply libraries/app/include/graphene/app/database_api.hpp]: fixed at https://github.com/bitshares/bitshares-core/pull/1749/commits/0968f95c2e8b761808bc3058ebfae44a21814dbc

File: libraries/chain/include/graphene/chain/asset_object.hpp (hunk 18 above)
Comment (oxarbitrage) [reply libraries/chain/include/graphene/chain/asset_object.hpp]: correct, thank you https://github.com/bitshares/bitshares-core/pull/1749/commits/89949957341c91acfa16053d90870bfb9a49dbea

File: libraries/chain/include/graphene/chain/market_object.hpp (hunk 15 above)
Comment (oxarbitrage) [reply libraries/chain/include/graphene/chain/market_object.hpp]: ok, changed at https://github.com/bitshares/bitshares-core/pull/1749/commits/976b56bbe3fac494490ef57277e11674c1a8e597

File: libraries/app/database_api.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (oxarbitrage) [reply libraries/app/database_api.cpp]: this was resolved by https://github.com/bitshares/bitshares-core/commit/bda09a0989d8b5c030792b1dc6ecc12217b43868

**Comment by oxarbitrage** [comment]
@abitmore let me know what do you think about this approach to make `start` parameter optional in htlc calls: https://github.com/bitshares/bitshares-core/pull/1749/commits/d07c3745a7aa2e5879f15af634e2e850c5df44a0

Basically it adds the optional to the `database_api` calls while leave as it is the `database_api_impl` ones as it is. 

Thank you.

**Comment by abitmore** [comment]
@oxarbitrage `optional` parameters need to be put in the end of the parameter list. If our old APIs have a `start` already in the middle, leave it unchanged so far. Actually, we can even change `limit` to optional, better do it in a new ticket though, since lots to be discussed and done.

**Comment by oxarbitrage** [comment]
I honestly don't like that some calls will have `limit` at the end and others in the middle. I will prefer to keep the order `start`, `limit` in this pull specially if we plan to make them both optional in a new ticket.

File: libraries/app/application.cpp (hunk 2 above)
Comment (oxarbitrage) [reply libraries/app/application.cpp]: @sschiessl-bcp i opened a new ticket to discuss and implement limit and pagination for the mentioned api calls: https://github.com/bitshares/bitshares-core/issues/1753

what do we think about the 10 limit by default for `get_full_accounts`?

Comment (abitmore) [reply libraries/app/application.cpp]: @oxarbitrage: @sschiessl-bcp has mentioned elsewhere that the reference wallet only fetches one account every time with `get_full_accounts`.

Comment (oxarbitrage) [reply libraries/app/application.cpp]: yes, i agree with 10.

**Comment by sschiessl-bcp** [comment]
I find mass fetching full accounts not feasible anyways. If you could 
choose what details a I are interested in (mass fetching for HTLC 
overview), that would be something else.

Am 11. Mai 2019 00:46:42 schrieb Abit <notifications@github.com>:

**Comment by sschiessl-bcp** [comment]
Limit 10 for? Amount of accounts? Or amount of objects in the account details?

Am 11. Mai 2019 12:37:17 vorm. schrieb Alfredo Garcia 
<notifications@github.com>:

**Comment by abitmore** [comment]
> mass fetching for HTLC overview

@sschiessl-bcp I thought we've added an API to fetch all HTLCs, but it turns out that we haven't. Reopened #1713.By the way you can always use `get_objects` to massively fetch details.

**Comment by pmconrad** [comment]
> Limit 10 for?

Number of accounts per call.

The number of entries per sub-list is separately configurable and defaults to 100.

**Comment by sschiessl-bcp** [comment]
> > mass fetching for HTLC overview
> 
> @sschiessl-bcp I thought we've added an API to fetch all HTLCs, but it turns out that we haven't. Reopened #1713.By the way you can always use `get_objects` to massively fetch details.

Thanks, that will be helpfull to create HTLC order books!

My comment also referred to how `get_full_accounts` functions. If it fetches only one account, all details are reasonable. If fetching multiple accounts, I could only think of use-cases where you are interested in a specific feature, e.g. fetching 10 accounts, and all those 10 accounts have complete HTLC details filled. Mass fetching of 10 accounts, with all details filled, sounds unreasonable to me.

**Comment by oxarbitrage** [comment]
@sschiessl-bcp here is how `get_full_accounts` will work with this changes:

- Query a total of 10 accounts max by default in 1 call, before this pull request this was unlimited. We can limit to just 1 but we think 10 is reasonable from a performance point of view.
- The lists inside the get_full_accounts responses(balances, assets, htlcs, proposals, withdraws, orders, etc) will also have a limit(100 by default) of records to return.
-  For each list with limit, there is an alternative api call to get the full thing with pagination. For example, if an account haves more than 100 call orders the client can call `get_call_orders_by_account` to get the full list.

Let me know what do you think.

**Comment by sschiessl-bcp** [comment]
> @sschiessl-bcp here is how `get_full_accounts` will work with this changes:
> 
> * Query a total of 10 accounts max by default in 1 call, before this pull request this was unlimited. We can limit to just 1 but we think 10 is reasonable from a performance point of view.
> * The lists inside the get_full_accounts responses(balances, assets, htlcs, proposals, withdraws, orders, etc) will also have a limit(100 by default) of records to return.
> * For each list with limit, there is an alternative api call to get the full thing with pagination. For example, if an account haves more than 100 call orders the client can call `get_call_orders_by_account` to get the full list.
> 
> Let me know what do you think.

Is there a way to e.g. fetch those maximum 10 accounts and only get the HTLC sublists? Or do I always get all (up to 100 reach)?

**Comment by abitmore** [comment]
@sschiessl-bcp after bumped FC, when we have https://github.com/bitshares/bitshares-fc/pull/126 in bitshares-core, we'll review all APIs and make them more flexible, it may need time. By now, I think you need to call `get_htlc_by_from` one time per account.

**Comment by sschiessl-bcp** [comment]
@oxarbitrage 

What is the default sorting of the returned objects? Just by id ascending?
